presto-main/src/main/java/io/prestosql/server/security/jwt/JwkService.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.server.security.jwt;
+
+import com.google.common.io.Closer;
+import io.airlift.concurrent.Threads;
+import io.airlift.http.client.HttpClient;
+import io.airlift.http.client.Request;
+import io.airlift.http.client.StringResponseHandler.StringResponse;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.annotation.processing.Generated;
+
+import java.io.IOException;
+import java.net.URI;
+import java.security.PublicKey;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static io.airlift.http.client.Request.Builder.prepareGet;
+import static io.airlift.http.client.StringResponseHandler.createStringResponseHandler;
+import static io.prestosql.server.security.jwt.JwkDecoder.decodeKeys;
+import static java.util.Objects.requireNonNull;
+
+public final class JwkService
+{
+    private static final Logger log = Logger.get(JwkService.class);
+
+    private final URI address;
+    private final HttpClient httpClient;
+    private final Duration refreshDelay;
+    private final AtomicReference<Map<String, PublicKey>> keys;
+
+    @Generated("this")
+    private Closer closer;
+
+    public JwkService(URI address, HttpClient httpClient, Duration refreshDelay)
+    {
+        this.address = requireNonNull(address, "address is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+        this.refreshDelay = requireNonNull(refreshDelay, "refreshDelay is null");
+
+        this.keys = new AtomicReference<>(fetchKeys());
+    }
+
+    @PostConstruct
+    public synchronized void start()
+    {
+        if (closer != null) {
```

Comment:
illegal state?

Reply:
Typically our start methods are idempotent, so if already started then it is "ok"